ai-integrations/langgraph.ipynb

```diff
@@ -0,0 +1,598 @@
+    "        result.append(ToolMessage(content=observation, tool_call_id=tool_call[\"id\"]))\n",
+    "    # Write `result` to the `messages` attribute of the graph state\n",
+    "    return {\"messages\": result}\n",
+    "\n",
+    "# The `tools` node should run the `tools_node` function\n",
+    "graph.add_node(\"tools\", tools_node)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from langgraph.graph import END\n",
+    "\n",
+    "# Add normal edges using the `add_edge` method\n",
+    "# Add an edge from the START node to the `agent` node\n",
+    "graph.add_edge(START, \"agent\")\n",
+    "\n",
+    "# Add an edge from the `tools` node to the `agent` node\n",
+    "graph.add_edge(\"tools\", \"agent\")\n",
+    "\n",
+    "# Define a conditional edge\n",
+    "def route_tools(state: GraphState):\n",
+    "    \"\"\"\n",
+    "    Uses a conditional_edge to route to the tools node if the last message\n",
+    "    has tool calls. Otherwise, route to the end.\n",
+    "    \"\"\"\n",
+    "    # Get messages from graph state\n",
+    "    messages = state.get(\"messages\", [])\n",
+    "    if len(messages) > 0:\n",
+    "        # Get the last AI message from messages\n",
+    "        ai_message = messages[-1]\n",
+    "    else:\n",
+    "        raise ValueError(f\"No messages found in input state to tool_edge: {state}\")\n",
+    "    \n",
+    "    # Check if the last message has tool calls\n",
+    "    if hasattr(ai_message, \"tool_calls\") and len(ai_message.tool_calls) > 0:\n",
+    "        # If yes, return \"tools\"\n",
+    "        return \"tools\"\n",
+    "    # If no, return END\n",
+    "    return END\n",
+    "\n",
+    "# Use the `add_conditional_edges` method to add a conditional edge from the `agent` node to the `tools` node\n",
+    "# based on the output of the `route_tools` function\n",
+    "graph.add_conditional_edges(\n",
+    "    \"agent\",\n",
+    "    route_tools,\n",
+    "    {\"tools\": \"tools\", END: END},\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from IPython.display import Image, display\n",
+    "\n",
+    "# Compile the graph\n",
+    "app = graph.compile()\n",
+    "\n",
+    "# Optionally, visualize the graph\n",
+    "try:\n",
+    "    display(Image(app.get_graph().draw_mermaid_png()))\n",
+    "except Exception:\n",
+    "    pass"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Execute the graph"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Stream outputs from the graph as they pass through its nodes\n",
+    "def execute_graph(user_input: str) -> None:\n",
+    "    \"\"\"\n",
+    "    Stream outputs from the graph\n",
+    "\n",
+    "    Args:\n",
```

Comment:
Similar nit re: args here - especially after we omitted this from the function documentation above, which _does_ also have args.